common/config_manager.py

```diff
@@ -1,242 +1,245 @@
-        return attribute
-    else:
-        try:
-            if not connection or not cursor:
-                connection, cursor = quick_db_connect()
-
-            query = "SELECT value FROM settings WHERE name = %s"
-            cursor.execute(query, (attribute_name,))
-            row = cursor.fetchone()
-
-            if not keep_connection_open:
-                connection.close()
-
-            value = row.get("value", None) if row else None
-            if not raw and value is not None:
-                value = json.loads(value)
-        except (Exception, psycopg2.DatabaseError) as error:
-            raise ConfigException("Error getting setting {}: {}".format(attribute_name, repr(error)))
-
-        finally:
-            if connection is not None and not keep_connection_open:
-                connection.close()
-
-        if value is None:
-            # no value explicitly defined in the database...
-            if default is not None:
-                # if an explicit default was provided, return it
-                return default
-            elif attribute_name in config_definition and "default" in config_definition[attribute_name]:
-                # if a default is available from the config definition, return that
-                return config_definition[attribute_name]["default"]
-            else:
-                # use None as the last resort
+    db = None
+    dbconn = None
+    cache = {}
+
+    core_settings = {}
+    config_definition = {}
+    tag_context = []  # todo
+
+    def __init__(self, db=None):
+        # ensure core settings (including database config) are loaded
+        self.load_core_settings()
+        self.load_user_settings()
+
+        # establish database connection if none available
+        self.db = db
+
+    def with_db(self, db=None):
+        """
+        Initialise database
+
+        Not done on init, because something may need core settings before the
+        database can be initialised
+
+        :param db:  Database object. If None, initialise it using the core config
+        """
+        self.db = db if db else Database(logger=None, dbname=self.get("DB_NAME"), user=self.get("DB_USER"),
+                           password=self.get("DB_PASSWORD"), host=self.get("DB_HOST"),
+                           port=self.get("DB_PORT"), appname="config-reader") if not db else db
+
+    def load_user_settings(self):
+        """
+        Load settings configurable by the user
+
+        Does not load the settings themselves, but rather the definition so
+        values can be validated, etc
+        """
+        # basic 4CAT settings
+        self.config_definition.update(config_definition)
+
+        # module settings can't be loaded directly because modules need the
+        # config manager to load, so that becomes circular
+        # instead, this is cached on startup and then loaded here
+        module_config_path = self.get("PATH_ROOT").joinpath("backend/module_config.json")
```

Comment:
I'm cannot find where the `module_config.json` file is created. But as far as I can tell, the `config_definition` only comes into play when there is no definition in the database and no default value has been provided to `get`.

Reply:
Work in progress ;)